sql/core/src/main/scala/org/apache/spark/sql/execution/stat/FrequentItems.scala

```diff
@@ -117,10 +112,16 @@ object FrequentItems extends Logging {
     )
     val justItems = freqItems.map(m => m.baseMap.keys.toArray)
     val resultRow = Row(justItems : _*)
-    // append frequent Items to the column name for easy debugging
-    val outputCols = colInfo.map { v =>
-      StructField(v._1 + "_freqItems", ArrayType(v._2, false))
-    }
+
+    val originalSchema = df.schema
+    val outputCols = cols.map { name =>
+      val index = originalSchema.fieldIndex(name)
+      val originalField = originalSchema.fields(index)
+
+      // append frequent Items to the column name for easy debugging
+      StructField(name + "_freqItems", ArrayType(originalField.dataType, originalField.nullable))
+    }.toArray
+
```

Comment:
We can simply fix like this? https://github.com/apache/spark/compare/master...maropu:SPARK-28818

Reply:
We can, but I felt what the code was doing already was a little weird. Like is there a good reason to split out the creation of `colInfo` and `outputCols`? Additionally it seems more readable to me to actually name the variable rather than use `_*` and `_1`, `_2`?

Reply:
In the interests of leaving the code in a better state than I found it 😄